File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-import-type-side-effects.test.ts
import { noFormat, RuleTester } from '@typescript-eslint/rule-tester';
import { getFixturesRootDir } from '../RuleTester.ts';

const rootPath = getFixturesRootDir();

const ruleTester = new RuleTester({
  // @ts-ignore
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootPath,
    },
  },
});

ruleTester.run('no-import-type-side-effects', {
  valid: [
    "import T from 'mod';",
    "import * as T from 'mod';",
    "import { T } from 'mod';",
    "import type { T } from 'mod';",
    "import type { T, U } from 'mod';",
    "import { type T, U } from 'mod';",
    "import { T, type U } from 'mod';",
    "import type T from 'mod';",
    "import type T, { U } from 'mod';",
    "import T, { type U } from 'mod';",
    "import type * as T from 'mod';",
    "import 'mod';",
  ],
  invalid: [
    {
      code: "import { type A } from 'mod';",
      errors: [{ messageId: 'useTopLevelQualifier' }],
      output: "import type { A } from 'mod';",
    },
    {
      code: "import { type A as AA } from 'mod';",
      errors: [{ messageId: 'useTopLevelQualifier' }],
      output: "import type { A as AA } from 'mod';",
    },
    {
      code: "import { type A, type B } from 'mod';",
      errors: [{ messageId: 'useTopLevelQualifier' }],
      output: "import type { A, B } from 'mod';",
    },
    {
      code: "import { type A as AA, type B as BB } from 'mod';",
      errors: [{ messageId: 'useTopLevelQualifier' }],
      output: "import type { A as AA, B as BB } from 'mod';",
    },
  ],
});

File: internal/rules/no_import_type_side_effects/no_import_type_side_effects_test.go
package no_import_type_side_effects

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/rule_tester"
	"github.com/web-infra-dev/rslint/internal/rules/fixtures"
)

func TestNoImportTypeSideEffectsRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoImportTypeSideEffectsRule, []rule_tester.ValidTestCase{
		// Valid cases
		{Code: "import T from 'mod';"},
		{Code: "import * as T from 'mod';"},
		{Code: "import { T } from 'mod';"},
		{Code: "import type { T } from 'mod';"},
		{Code: "import type { T, U } from 'mod';"},
		{Code: "import { type T, U } from 'mod';"},
		{Code: "import { T, type U } from 'mod';"},
		{Code: "import type T from 'mod';"},
		{Code: "import type T, { U } from 'mod';"},
		{Code: "import T, { type U } from 'mod';"},
		{Code: "import type * as T from 'mod';"},
		{Code: "import 'mod';"},
	}, []rule_tester.InvalidTestCase{
		// Invalid cases
		{
			Code:   "import { type A } from 'mod';",
			Output: []string{"import type { A } from 'mod';"},
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "useTopLevelQualifier"},
			},
		},
		{
			Code:   "import { type A as AA } from 'mod';",
			Output: []string{"import type { A as AA } from 'mod';"},
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "useTopLevelQualifier"},
			},
		},
		{
			Code:   "import { type A, type B } from 'mod';",
			Output: []string{"import type { A, B } from 'mod';"},
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "useTopLevelQualifier"},
			},
		},
		{
			Code:   "import { type A as AA, type B as BB } from 'mod';",
			Output: []string{"import type { A as AA, B as BB } from 'mod';"},
			Errors: []rule_tester.InvalidTestCaseError{
				{MessageId: "useTopLevelQualifier"},
			},
		},
	})
}
